hydrolib/core/basemodel.py

```diff
@@ -0,0 +1,127 @@
+"""
+Here we define our Pydantic `BaseModel` with custom settings,
+as well as a `FileModel` that inherits from a `BaseModel` but
+also represents a file on disk.
+
+"""
+from abc import ABC, abstractclassmethod
+from pathlib import Path
+from typing import Any, Callable, Optional, Type
+from warnings import warn
+
+from pydantic import BaseModel as PydanticBaseModel
+
+from hydrolib.core.io.base import DummmyParser, DummySerializer
+
+
+class BaseModel(PydanticBaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+        validate_assignment = True
+        use_enum_values = True
+        extra = "allow"
+
+
+class FileModel(BaseModel, ABC):
+    """Base class to represent models with a file representation.
+
+    It therefore always has a `filepath` and if it is given on
+    initilization, it will parse that file.
+
+    This class extends the `validate` option of Pydantic,
+    so when when a Path is given to a field with type `FileModel`,
+    it doesn't error, but actually initializes the `FileModel`.
+    """
+
+    filepath: Optional[Path] = None
+
+    def __init__(self, filepath: Optional[Path] = None, *args, **kwargs):
+        """Initialize a model.
+
+        The model is empty (with defaults) if no `filepath` is given,
+        otherwise the file at `filepath` will be parsed."""
+        # Parse the file if path is given
+        if filepath:
+            data = self._load(filepath)
+            data["filepath"] = filepath
+            kwargs.update(data)
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def validate(cls: Type["FileModel"], value: Any) -> "FileModel":
+        # Enable initialization with a Path.
+        if isinstance(value, Path):
+            # Pydantic Model init requires a dict
+            value = {"filepath": value}
+        return super().validate(value)
+
+    def _load(self, filepath: Path):
+        if Path(filepath).is_file():
+            return self._parse(filepath)
+        else:
+            warn(f"File: `{filepath}` not found, skipped parsing.")
+            return {}
+
+    def save(self, folder: Path, force=False):
+        """Save model and child models to their set filepaths.
+
+        For models with an unset filepath, we generate one based
+        on the given `folder`.
+
+        If `force` is set, we override the folder part of
+        already set filepaths, which can be used to copy complete models.
+        """
+
+        if not self.filepath:
+            self.filepath = folder / self._generate_name()
+
+        if force:
+            self.filepath = folder / self.filepath.name
+
+        # Convert child FileModels first
+        exclude = {"filepath"}
+        filemodel_fields = {}
+        for name, value in self:
+            if isinstance(value, FileModel):
+                filepath = value.save(folder, force)
+                filemodel_fields[name] = filepath
+                exclude.add(name)
```

Comment:
Does this prevent that the reference to this child file will not be written to the parent file? For example, in the mdu file we reference the network file.

Reply:
It prevents the child file being returned as {fielda: value, etc} (automatic due to Pydantic), now it is returned as Path("location/of/child") by ourselves.